Cok 2542 slice - Specifications (#3426)
dafny-lang/dafny: docs/DafnyRef/Specifications.md

There's (unfortunately) no form of it for array
elements -- but to account for unchanged elements, you can always write
`forall i | 0 <= i < |a| :: unchanged(a[i])`.

A ``FrameField`` is not taken into consideration for
lambda expressions.

### 19.1.5. Reads Clause ([grammar](#g-reads-clause)) {#sec-reads-clause}

Examples:
<!-- %no-check -->
```dafny
const o: object
const o, oo: object
function f()
  reads *
function g()
  reads o, oo
function h()
  reads { o }
```

Functions are not allowed to have side effects; they may also be restricted in

> what they can read. The _reading frame_ of a function (or predicate) is all

Reviewer: ```suggestion
what they can read. The _reading frame_ of a function (or predicate) consists of all
```

set x: int, o: object | o in F(x) :: o
```

For each function value `f`, Dafny defines the function `f.reads`,
which takes the same arguments as `f` and returns that set of objects
that `f` reads (according to its reads clause) with those arguments.

> `f.reads` has type `int ~> set<object>`.

Reviewer: Are you referring to a particular function `f` here?

Author: No, its 'forall' functions f

Reviewer: Then, why do you write `f.reads` has type `int ~> set<object>`. Shouldn't it be, say, `T` instead of `int`?

Author: f is a function,. f.reads is also a function. It takes the same arguments as f, but returns the set of objects read. Similarly f.requires takes the same arguments as f but returns a boolean (wither the precondition is satisfied or not).

The use case is to be able to include, say f.requires in a precondition of another funciton/method that has f as an argument.  So map(f: T->bool, s: seq<T>): seq<bool> might have forall t: t in s :: f.requires(t) as a precondition.

Author: Change the int to T

Note, only `reads` clauses, not `modifies` clauses, are allowed to
include functions as just described.

### 19.1.6. Modifies Clause ([grammar](#g-modifies-clause)) {#sec-modifies-clause}

Examples:
<!-- %check-resolve -->
```dafny
class A { var f: int }
const o: object?
const p: A?
method M()
  modifies { o, p }
method N()
  modifies { }
method Q()

>   modifies o,p`f

Reviewer: ```suggestion
  modifies o, p`f
```

ACSL and ACSL++ use the `assigns` keyword, but with _modify_ (b) semantics.
Ada/SPARK's dataflow contracts encode _write_ (a) semantics.

### 19.1.7. Invariant Clause ([grammar](#g-invariant-clause)) {#sec-invariant-clause}

Examples:
<!-- %check-resolve-warn Specifications.2.expect -->
```dafny
method m()
{
  var i := 10;
  while 0 < i
    invariant 0 <= i < 10
}
```

An `invariant` clause is used to specify an invariant
for a loop. If more than one `invariant` clause is given for

> a loop the effective invariant is the conjunction of

Reviewer: ```suggestion
a loop, the effective invariant is the conjunction of
```

> the conditions specified, in their textual order.

Reviewer: What is a textual order?

Author: The order in the source text, in case previous ones are needed to establish later ones or to ensure the later ones are well-defined. Same as for preconditions and for postconditions

Reviewer: Yes, thank you.

The invariant must hold on entry to the loop. And assuming it
is valid on entry to a particular iteration of the loop, 
Dafny must be able to prove that it then
holds at the end of that iteration of the loop.

## 19.2. Method Specification ([grammar](#g-method-specification)) {#sec-method-specification}

Examples:
<!-- %check-resolve -->
```dafny
class C {
  var next: C?
  var value: int

  method M(i: int) returns (r: int)
    requires i >= 0
    modifies next
    decreases i
    ensures r >= 0
  { 
    ... 
  }
}
```

> A method specification is zero or more `modifies`, `requires`,

Reviewer: ```suggestion
A method specification consists of zero or more `modifies`, `requires`,
```

Reviewer: This sounds quite _formal_ to me. Same elsewhere. Is this a standard way of expressing these sort of things?

Author: I'd say so. This document is a combination of a reference manual (for which formality is typical), a user guide and a tutorial. So the language style is a bit of a mix.

An iterator specification applies both to the iterator's constructor
method and to its `MoveNext` method.
- The `reads` and `modifies`
clauses apply to both of them. 
- The `requires` and `ensures` clauses apply to the constructor.
- The `yield requires` and `yield ensures` clauses apply to the `MoveNext` method.

Examples of iterators, including iterator specifications, are given in
[Section 15](#sec-iterator-types). Briefly
- a requires clause gives a precondition for creating an iterator
- an ensures clause gives a postcondition when the iterator exits (after all iterations are complete)
- a decreases clause is used to show that the iterator will eventually terminate
- a yield requires clause is a precondition for calling `MoveNext`
- a yield ensures clause is a postcondition for calling `MoveNext`

## 19.6. Loop Specification ([grammar](#g-loop-specification)) {#sec-loop-specification}

A loop specification provides the information Dafny needs to
prove properties of a loop. It contains `invariant`,
`decreases`, and `modifies` clauses.

The `invariant` clause
is effectively a precondition and it along with the
negation of the loop test condition provides the postcondition.

> The `decreaases` clause is used to prove termination.

Reviewer: ```suggestion
The `decreases` clause is used to prove termination.
```
